Improve and Expand API Documentation

omad opened the pull request:
### Reason for this pull request

The existing API documentation is both really hard to browse, and is missing lots and lots of valuable functionality.


### Proposed changes
- Flatten out the structure of the current API docs so that everything is listed on a single `Ctrl-F` able page.
- Add lots of currently unmentioned functions/methods and classes into the front API docs page.

See [proposed new API page](https://damien-agdc.readthedocs.io/en/latest/dev/api/index.html) for a view of the new page. Compare with [current API page](https://datacube-core.readthedocs.io/en/latest/dev/api/index.html) 



 - ~[ ] Closes #xxxx~
 - ~[ ] Tests added / passed~
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

<!--
See https://github.com/blog/2111-issue-and-pull-request-templates for more
information on pull request templates.

-->

omad committed:
Fix documentation build warnings

omad committed:
More API documentation improvements

omad committed:
API doc fixes

omad committed:
API doc fixes

omad committed:
Add what's new comment about API docs

mpaget commented:
Looks good, @omad. Nice initiative, makes browsing the scope and discovering specific functions much easier.

Direct links to each function is helpful for more information. A spot check suggests this is done for most functions but is missing (if applicable) for:
- Product Querying
- Product Addition/Modification
And it might be helpful to have corresponding text descriptions for the "Creating Geometries" functions.

Also, it would be good to do something similar (flatten and incorporate into this page) for the "Everything Else"s but they can be a separate pull request.

Kirill888 commented on docs/dev/api/index.rst:
```diff
@@ -1,18 +1,351 @@
+---------------
+
+When connected to an ODC Database, these methods are available for
+adding, updating and deleting datasets::
```
except we actually can not **delete** datasets, only **archive** datasets.

Kirill888 commented on docs/dev/api/index.rst:
```diff
@@ -1,18 +1,351 @@
+   :nosignatures:
+   :toctree: generate/
+
+   utils.geometry.BoundingBox
```
`BoundingBox` is not "CRS aware", and it's really annoying that it is the odd one out.

Kirill888 reviewed (dismissed):
Looks good, see some minor notes below. Also there are still a lot of places where we refer to "products" as "dataset types", haven't seen them in the diff but there are a few in the "Product Querying" section, I guess they get pulled from python doc strings.

andrewdhicks reviewed (dismissed):
:+1:

omad committed:
API Docs updates

- Docstrings for geometry constructors
- Fix links to Product methods
- Pull Index/index_connect() into main API list
-
